dowhy/causal_identifier/backdoor_identifier.py

```diff
@@ -36,119 +22,160 @@ class CausalIdentifier:
     BACKDOOR_EFFICIENT = "efficient-adjustment"
     BACKDOOR_MIN_EFFICIENT = "efficient-minimal-adjustment"
     BACKDOOR_MINCOST_EFFICIENT = "efficient-mincost-adjustment"
+
+
+class BackdoorIdentifier:
+    """Class that implements different identification methods.
+
+    Currently supports backdoor and instrumental variable identification methods. The identification is based on the causal graph provided.
+
+    """
+
+    MAX_BACKDOOR_ITERATIONS = 100000
+
     METHOD_NAMES = {
-        BACKDOOR_DEFAULT,
-        BACKDOOR_EXHAUSTIVE,
-        BACKDOOR_MIN,
-        BACKDOOR_MAX,
-        BACKDOOR_EFFICIENT,
-        BACKDOOR_MIN_EFFICIENT,
-        BACKDOOR_MINCOST_EFFICIENT,
+        BackdoorAdjustmentMethod.BACKDOOR_DEFAULT,
+        BackdoorAdjustmentMethod.BACKDOOR_EXHAUSTIVE,
+        BackdoorAdjustmentMethod.BACKDOOR_MIN,
+        BackdoorAdjustmentMethod.BACKDOOR_MAX,
+        BackdoorAdjustmentMethod.BACKDOOR_EFFICIENT,
+        BackdoorAdjustmentMethod.BACKDOOR_MIN_EFFICIENT,
+        BackdoorAdjustmentMethod.BACKDOOR_MINCOST_EFFICIENT,
     }
     EFFICIENT_METHODS = {
-        BACKDOOR_EFFICIENT,
-        BACKDOOR_MIN_EFFICIENT,
-        BACKDOOR_MINCOST_EFFICIENT,
+        BackdoorAdjustmentMethod.BACKDOOR_EFFICIENT,
+        BackdoorAdjustmentMethod.BACKDOOR_MIN_EFFICIENT,
+        BackdoorAdjustmentMethod.BACKDOOR_MINCOST_EFFICIENT,
     }
-    DEFAULT_BACKDOOR_METHOD = BACKDOOR_DEFAULT
+    DEFAULT_BACKDOOR_METHOD = BackdoorAdjustmentMethod.BACKDOOR_DEFAULT
 
     def __init__(
         self,
-        graph,
-        estimand_type,
-        method_name="default",
-        proceed_when_unidentifiable=False,
+        estimand_type: CausalIdentifierEstimandType,
+        backdoor_adjustment: BackdoorAdjustmentMethod = BackdoorAdjustmentMethod.BACKDOOR_DEFAULT,
+        proceed_when_unidentifiable: bool = False,
+        optimize_backdoor: bool = False,
+        costs: Optional[List] = None,
```

Comment:
Should costs be a part of the constructor or an argument to identify_effect?

Reply:
I made that question myself many times during the refactor haha. What do you think? I was thinkng that the `__init__()` parameters should be configuration used in the `identify_effect()` and the parameters to `identify_effect()` the data to make the calculation... I wasn't so sure in which one `costs` would go.

Reply:
@esmucler based on the scenarios you've seen, do you have thoughts about whether the costs passed into the efficient backdoor algorithm are more rightly considered to be configuration parameters vs data parameters?

background: This is coming up in the context of a refactoring of the DoWhy to have a more functional, less object oriented, API

Reply:
Hi folks. I think it makes more sense to consider them as configuration parameters for the constructor, or even as attributes of the graph, but I guess the latter is out of the question for now. For one, I can't imagine atm why a user would want to call identify_effect() on the same setup for different cost values.

Reply:
I'll leave it as constructor params for now, thanks for your input @esmucler